decisions/migrations.md

```diff
@@ -0,0 +1,262 @@
+# Migrations
+
+- *Date*: 2022-10-19
+- *Deciders*:
+- *Status*: DRAFT
+- *Outcome*: 
+- *Related Material*: [Database migration Issue](https://github.com/openmsupply/open-msupply/issues/704)
+
+## Requirements
+
+1. Update schema without re-initialising database
+2. Migrate data
+   a. Before and/or after schema update
+   b. Without schema update (just data migration)
+3. Do above only for the migrations that are applicable to from current database version to current app version (don't re-run old migrations)
+4. Record migration results
+5. Easy way to explore full schema (all of the migrations applied, for a particular version)
+6. Test data migrations
+7. Should work with RC versions
+8. (maybe) Consolidate common sql for PG/Sqlite (we have quite a few duplicate sql statements)
+9. (maybe) Use strongly typed diesel db if possible, for data migrations
+10. Migration code needs to run in isolation (no production code should run in parallel)
+
+## Examples/Extra
+
+### 2.a 
+
+There might be a need to do multiple schema and data updates within one version upgrade. For example if a required field is added that needs to be populated with some value derived from existing database values, we would want to:
+* add schema without the constraint
+* add the values 
+* add the constraint
+
+Another example is when sync buffer was consolidated:
+* Add new sync buffer
+* Copy rows from remote and central to the new sync buffer (see proposed solution, that touches on `9.`)
+* Remove old sync buffer
+
+Above can also be done by keeping existing sync buffer in memory than doing one schema migration with adding and removing sync buffer then populating new sync buffer from memory. 
+
+I think it's better to keep migrations flexible, without a particular pattern of operation sequence.
+
+### 2.b
+
+If we add a new field/table in central and remote is not up to date, the field/table will still be synced and recorded in sync buffer. And when we update to a new version of remote that has this new field/table we can just crawl through sync buffer and try to re-integrate this new field/table. This avoids needing to re-sync this info, and we can avoid implementing this mechanism
+
+Remote data should be safe with above example, as central shouldn't be updating foreign data.
+
+### 4.
+
+General data about migration result should be logged for debugging purposes, and on migration failure we can't guarantee correct behaviour of the app, so we would need to either disable server and not allow it to start again or revert to previous data version and inform the admin/user that app can be downgraded. 
+
+I think ability to revert to previous version (on failure) allows for system to be available to the users in timely manner, in case errors in migration occur.
```

Comment:
👍  plus one on that.
run the entire migration in a transaction? 
backup database before starting?
when upgrading the server, keep a copy of the previous binary to enable an easy rollback?

Reply:
Yeah, very close to what I was thinking, I am hoping to push this to cary over, get the base migrations in and then strategise about rollback, it seems that `option 3` would allow quite a bit of flexibility ad control to go with either approach. I've added a comment about android rollback 9aea1d161a98d7450ee6e9d81501c99e314f2b47 and added this to one of the task in epic https://github.com/openmsupply/open-msupply/issues/704